pkgs/development/python-modules/pypdfium2/bin.nix

```diff
@@ -0,0 +1,147 @@
+  pdfiumVersion_build = lib.elemAt parsedNumericVersionComponents 2;
+  pdfiumVersion_patch = lib.elemAt parsedNumericVersionComponents 3;
+
+  # Contains more than a single dir; `builtins.fetchTarball` and `fetchzip` cannot handle that yet,
+  # so we need to manually unpack it below.
+  pdfiumPrebuiltTar = (fetchurl {
+    url = "https://github.com/bblanchon/pdfium-binaries/releases/download/chromium/${toString pdfiumVersion_build}/pdfium-${pdfium.platformFileInfix}.tgz";
+    hash = pdfium.hash;
+  });
+
+  pdfium-bin = runCommand "pdfium-bin" {} ''
+    mkdir "$out"
+    tar xf "${pdfiumPrebuiltTar}" --directory "$out"
+  '';
+
+  ctypesgen_pypdfium_fork_package =
+    { lib
+    , buildPythonApplication
+    , fetchFromGitHub
+    , toml
+    , setuptools-scm
+    }:
+      buildPythonApplication rec {
+        pname = "ctypesgen";
+        version = "3a811c28160b7c742f97ee50bfe55b18d4021ca0";
+
+        src = fetchFromGitHub {
+          owner = "pypdfium2-team"; # fork needed for `pypdfium2`
+          repo = "ctypesgen";
+          rev = version;
+          hash = "sha256-VqZ7Xn/7mrBjzIA7RUQt4WkrOttJMb4XmFWPpGtPfHY=";
+        };
+
+        # This package uses setuptools-scm to derive the version from the git repo (which we don't have),
+        # so use this environment variable to set it manually instead.
+        SETUPTOOLS_SCM_PRETEND_VERSION = "1.1.1+g${version}"; # we use the most recent tag and append the git version
+
+        doCheck = false;
+
+        propagatedBuildInputs = [
+          toml
+          setuptools-scm
+        ];
+      };
+  ctypesgen_pypdfium_fork = callPackage ctypesgen_pypdfium_fork_package {};
+
+in buildPythonPackage rec {
+  pname = "pypdfium2";
+  version = "4.24.0";
+
+  src = fetchFromGitHub {
+    owner = "pypdfium2-team";
+    repo = "pypdfium2";
+    rev = version;
+    hash = "sha256-dH7IViCfcYuCLZAuqHkl03aXOeFEjm93zKiRraxGeGQ=";
+  };
+
+  # Following instructions from section
+  # > With caller-built data files (this is expected to work offline)
+  # https://github.com/pypdfium2-team/pypdfium2/tree/30c60af438b7cd90e22d42dd2ba5bffdeb568c42#install-source-caller
+  # with the addition of `--compile-libdirs`, which fixes `ctypesgen` warning:
+  #     WARNING: Could not load library 'pdfium'. Okay, I'll try to load it at runtime instead.
+  preConfigure = ''
+    "${ctypesgen_pypdfium_fork}/bin/ctypesgen" \
+      --library pdfium \
+      --compile-libdirs "${pdfium-bin}/lib" \
+      --runtime-libdirs "${pdfium-bin}/lib" \
+      --headers "${pdfium-bin}"/include/fpdf*.h \
+      -o src/pypdfium2_raw/bindings.py \
+      --strip-build-path=. \
+      --no-srcinfo
+
+    cat > src/pypdfium2_raw/version.json <<END
+    {
+      "major": ${toString pdfiumVersion_major},
+      "minor": ${toString pdfiumVersion_minor},
+      "build": ${toString pdfiumVersion_build},
+      "patch": ${toString pdfiumVersion_patch},
+      "n_commits": 0,
+      "hash": null,
+      "origin": "pdfium-binaries/nixos",
```

Comment:
It's quite unclear whether this should be `pdfium-binaries/nixos` or `system/nixos`, but that's our upstream fault. In theory `pdfium-binaries` implies bundling, but `system` doesn't include the meaning that the binary comes from pdfium-binaries.
Since a hashed, explicit runtime libdir is as good as bundling, I'd leave it as-is for now.
I plan to revise the version file (and how our code is writing it) to cleanly include all these meanings: library location (bundled/external), package flavor (pypa / conda / linux distro), and binary builder (pdfium-binaries / distro / other)